dspace-api/src/main/java/org/dspace/authenticate/OidcAuthenticationBean.java

```diff
@@ -0,0 +1,272 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(OidcAuthenticationBean.class);
+
+    private static final String OIDC_AUTHENTICATED = "oidc.authenticated";
+
+    @Autowired
+    private ConfigurationService configurationService;
+
+    @Autowired
+    private OidcClient oidcClient;
+
+    @Autowired
+    private EPersonService ePersonService;
+
+    @Override
+    public boolean allowSetPassword(Context context, HttpServletRequest request, String username) throws SQLException {
+        return false;
+    }
+
+    @Override
+    public boolean isImplicit() {
+        return false;
+    }
+
+    @Override
+    public boolean canSelfRegister(Context context, HttpServletRequest request, String username) throws SQLException {
+        return canSelfRegister();
+    }
+
+    @Override
+    public void initEPerson(Context context, HttpServletRequest request, EPerson eperson) throws SQLException {
+    }
+
+    @Override
+    public List<Group> getSpecialGroups(Context context, HttpServletRequest request) throws SQLException {
+        return List.of();
+    }
+
+    @Override
+    public String getName() {
+        return OIDC_AUTH_ATTRIBUTE;
+    }
+
+    @Override
+    public int authenticate(Context context, String username, String password, String realm, HttpServletRequest request)
+        throws SQLException {
+
+        if (request == null) {
+            LOGGER.warn("Unable to authenticate using OIDC because the request object is null.");
+            return BAD_ARGS;
+        }
+
+        if (request.getAttribute(OIDC_AUTH_ATTRIBUTE) == null) {
+            return NO_SUCH_USER;
+        }
+
+        String code = (String) request.getParameter("code");
+        if (StringUtils.isEmpty(code)) {
+            LOGGER.warn("The incoming request has not code parameter");
+            return NO_SUCH_USER;
+        }
+
+        return authenticateWithOidc(context, code, request);
+    }
+
+    private int authenticateWithOidc(Context context, String code, HttpServletRequest request) throws SQLException {
+
+        OidcTokenResponseDTO accessToken = getOidcAccessToken(code);
+        if (accessToken == null) {
+            LOGGER.warn("No access token retrieved by code");
+            return NO_SUCH_USER;
+        }
+
+        Map<String, Object> userInfo = getOidcUserInfo(accessToken.getAccessToken());
+
+        String email = getAttributeAsString(userInfo, getEmailAttribute());
+        if (StringUtils.isBlank(email)) {
+            LOGGER.warn("No email found in the user info attributes");
+            return NO_SUCH_USER;
+        }
+
+        EPerson ePerson = ePersonService.findByEmail(context, email);
+        if (ePerson != null) {
+            request.setAttribute(OIDC_AUTHENTICATED, true);
+            return ePerson.canLogIn() ? logInEPerson(context, ePerson) : BAD_ARGS;
+        }
+
+        return canSelfRegister() ? registerNewEPerson(context, userInfo, email) : NO_SUCH_USER;
```

Comment:
If canSelfRegister() returns false and no Eperson exists, this silently fails every single login.  Debugging that is a huge pain (I've been fighting this for a day now it seems).  We need to add a WARN or INFO logs here which can say something like "Self registration is currently disabled, but no EPerson could be found." or similar.